main/test/ca/uwaterloo/flix/library/TestLazyList.flix

```diff
@@ -255,4 +380,151 @@ namespace TestLazyListOld {
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // filter (impure)                                                         //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def testFilterImpure01(): Bool & Impure =
+        Nil |> LazyList.toLazy |> LazyList.filter(x -> x > 100 as & Impure) |> LazyList.toList == Nil
+
+    @test
+    def testFilterImpure02(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 0 as & Impure) |> LazyList.toList == 1 :: 2 :: 3 :: Nil
+
+    @test
+    def testFilterImpure03(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 1 as & Impure) |> LazyList.toList == 2 :: 3 :: Nil
+
+    @test
+    def testFilterImpure04(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 2 as & Impure) |> LazyList.toList == 3 :: Nil
+
+    @test
+    def testFilterImpure05(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 3 as & Impure) |> LazyList.toList == Nil
+
+    @test
+    def testFilterImpure06(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 100 as & Impure) |> LazyList.toList == Nil
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // filter filter                                                           //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def testFilterFilter01(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 1) |> LazyList.filter(x -> x < 3) |> LazyList.toList == 2 :: Nil
+
+    @test
+    def testFilterFilter02(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 1 as & Impure) |> LazyList.filter(x -> x < 3) |> LazyList.toList == 2 :: Nil
+
+    @test
+    def testFilterFilter03(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 1) |> LazyList.filter(x -> x < 3 as & Impure) |> LazyList.toList == 2 :: Nil
+
+    @test
+    def testFilterFilter04(): Bool & Impure =
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |> LazyList.filter(x -> x > 1 as & Impure) |> LazyList.filter(x -> x < 3 as & Impure) |> LazyList.toList == 2 :: Nil
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // filter filter fusion                                                    //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def testFilterFusion01(): Bool & Impure =
+        let l = ref Nil;
+        (1 :: 2 :: 3 :: Nil) |> LazyList.toLazy |>
+        LazyList.filter(_x -> {l := "a" :: deref l; true}) |>
+        LazyList.filter(_x -> {l := "b" :: deref l; true}) |>
+        LazyList.toList;
```

Comment:
Converting to a list would evaluate everything, whether pure or impure, right?

So to ensure the eagerness principles are working properly, we shouldn't convert to a list.

Reply:
Ah- you mean that the final call to `toList` should not be required, could be removed, and we should still see the same result? I buy that.